### DISCUSSION TITLE
Is there a way to make mg_mgr_poll() non blocking or make mg totally event drivent ?

### DISCUSSION THREAD

**Discussion by linuxguy123** [cesanta/mongoose#1773] [Ask questions, get answers] [answered]
I dislike having mg_mgr_poll() in my main program loop.  Is there a way to make mg_mgr_poll() non blocking or make mg totally event driven ?

Is there a minimum timeout for mg_mgr_poll() ? 5 ms ?  1 ms ?

Is there an event somewhere that could be responded to to initiate calling the mg_mgr_poll() process instead of having it in the main loop ?

Is there a way to make the Mongoose library totally event driven, totally bypassing the need to call mg_mgr_poll() ?     

Mongoose rocks.  Kudos to the devs.

**Comment by scaprile** [comment]
Mongoose is totally event driven, internally. `mg_mgr_poll()` is an event manager, it collects events and dispatches them to their respective handlers.
It blocks until either something arrives from the network or the timeout expires.
If your timeout is too short, then your loop time will be determined by the execution times instead of the blocking timeout, try it out and see if that is what you want. The drawback is that your handlers will receive a POLL event every time the timeout expires.

**Reply by scaprile** [reply]
> It blocks until either something arrives from the network or the timeout expires.

That is basically `select(socket, timeout)`; once you get something, it will react and call your handler.
Can you provide a use case and example code where the current behavior is not satisfactory ?

**Comment by linuxguy123** [comment, marked as answer]
I know that mg is event driven internally.  But it still needs the poll() call to respond to those events.

Is there a way to respond directly to an internal event without having to use poll() ?

I want to start using mg for some time sensitive UDP and WebSocket messaging.  Is there a way to respond directly to a message received event as soon as it happens ?

**Reply by cpq** [reply]
@linuxguy123 `mg_mgr_poll()` sleeps the requested time, but if there is any data arrives to any of the connections, the sleep gets interrupted. For example, if you do `mg_mgr_poll(&mgr, 1000)` - i.e. order mongoose to sleep 1 second, and a data arrives over a WebSocket connection 15 milliseconds after, then Mongoose gets interrupted right after data arrives - after 15 milliseconds (modulo OS scheduling time granularity). And, `mg_mgr_poll()` will return in 15ms, not in 1sec. And the response will be generated and sent straight away.

I think the timeout parameter for `mg_mgr_poll()` could be interpreted in some confusing ways.
That said, when you say "make mg_mgr_poll() totally event driven" - what API would you suggest? How do you see it?
